### PR TITLE
Feat save restore selections

### PR DESCRIPTION
**Pull request by geier1993**
Hey,

I noticed that you made the extension available finally. So I thought it is time to incorporate my efforts on saving and restoring selections. (The code was lounging around for some months...)

This feature also implements a kakoune-like behaviour: When switching to insert mode, selections are maintained (made visual using decorators) and restored when switching back to normal mode. So the user won't see a difference.

### CODE REVIEW

**Comment by 71** [comment]
Hey there,

Thanks for your PR! I'll look at it when I get the time (I'm rather busy these days), but in any case after quickly looking through the code, I'm sure some changes will have to be made to adhere to the coding style (e.g some spacing between `{` and `id`, using interfaces instead of constructors, doing casts with `<>`, no spacing between `if` and `(`).

In any case it seems like you did quite a lot of things, so I'll have to thoroughly look at it and understand what happens at some point before merging. :)

**Comment by geier1993** [comment]
> using interfaces instead of constructors
I don't know what you exactly mean. However you're right, you have to look through it.

Here is what happens:
* `DocumentHistory` in history.ts now stores marks and the last selections as `OffsetSelection`
* An OffsetSelection simply stores a range by offset (i.e. stand and end) instead by line and character. This makes it easier to handle changes with `TextDocumentContentChangeEvent`.
* The content of `lastSelections` is handled before every insertion. I had to make it explicit in every insert command in `commands/insert.ts` before the selections are modified for insertion.
* On every mode switch back to normal mode, the last selection is restored
* For the implementier of `OffsetRange` and `OffsetSelection` I tried to follow the original `Selection` and `Range` implementation.

Hope you like it. Took some time to implement this. It was the most performant and simple way I found to track changes.
The only thing I care about is how `document.offsetAt` is implemented. I have to use this function to map from back from `OffsetSelection` to `Selection` properly. If the implementation traverses the whole document in worst case, for large documents the performance will be poor. (I already worked with documents with 1M text lines).

Hope you like it

**Comment by 71** [comment]
Hey there,

Just wanted to chime in to say I haven't forgotten about this, I'm just still too busy to look at it. I'll try to find time to review it before I go on holidays, though! Thanks again for your work and for your explanation.

**Comment by 71** [comment]
There are still a few places where the coding style isn't right, but for now let's focus on more important things.

Seems like a lot is needed to make this work and I haven't grasped the implementation yet, but some things don't work as expected / like Kakoune.
- Moving the cursor after starting an insertion should remove the previous selection.
  - Select some text
  - Press A or I
  - Move the cursor using arrows
  - The previous selection remains (Kakoune behavior: the previous selection is cleared)
  - Re-enter normal mode
  - The previous selection is restored (Kakoune behavior: the current selection is kept)
- The above can also be observed by inserting a new line with `o`, for instance.
- Starting insertion at the beginning should _not_ include the selection.
  - Select some text
  - Press `i` to start insertion
  - Type some text
  - The typed text will be included in the selection (Kakoune behavior: the typed text is not included in the selection)

Congrats on making the `z` / `Z` commands work, it looks like it handles well non-trivial changes. I'll need to test this more anyway.

Since Dance is advertised as "alpha" quality, I don't mind merging your changes without them being 100% valid, as long as most of the things work. This will also help us get good bug reports. Would you be interested in doing that?

IIRC VS Code stores a lot of metadata, so `document.offsetAt` will likely have good performance (it certainly won't traverse the entire document, so at worst it's linear in the number of lines).

I noticed a few freezes, but haven't looked at where they come from yet.

**Comment by geier1993** [comment]
Sorry forget about that.
I try to fix the kakoune behaviour.

> Since Dance is advertised as "alpha" quality, I don't mind merging your changes without them being 100% valid, as long as most of the things work. This will also help us get good bug reports. Would you be interested in doing that?

What do you mean? I am definetly interested in merging the changes :)

**Comment by geier1993** [comment]
Okay the first problems remains. When a the cursor is moved in insert mode something goes wrong.
I don't know how to interact on selection changes in insert mode

**Comment by 71** [comment]
> Is there a way to interact with existing commands like "cursorLeft" "insertLineAbove"? The only thing I need is something like a "onInsertion" or "onCursorChange" callback in insertMode where I can reset the last selections the the document history in the extension.

Yep, in fact Dance [is already listening to selection changes](https://github.com/71/dance/blob/747fb49e99c93ad8055e39471aff34be8abe70e2/src/extension.ts#L229). You can similarly [listen to edit changes](https://code.visualstudio.com/api/references/vscode-api#workspace.onDidChangeTextDocument).

> What do you mean? I am definetly interested in merging the changes :)

I mean that it's okay if the feature has some hard-to-find edge cases. We'll just find them and fix them over time.

I like what you're doing here, and it sure looks like a hard problem. Thanks again for the help!

**Comment by geier1993** [comment]
Hey,

sorry for having not responded for a while. Had a lad off work...
The problem proved to be rather simple to solve. I just had to insert some lines to the Selection listener.
Fortunately the DocumentChanges Listener are called before the Selection Change listener. Hence the most secure way is to test whether the 'lastSelections' tracked in the documentHistory intersects with the new selection...

If the listener call order was different, the problem would be really ugly...

Have you found any other bugs so far?

PS: The merged the last master branch...

**Comment by geier1993** [comment]
Are there still any stiling issues? TSLint prints a lot of "must be braced" expressions which I ignored. However are stiling concers also covered by tslint?

File: src/extension.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import * as vscode from 'vscode'
 import { commands, Mode }      from './commands/index'
 import { HistoryManager }      from './history'
 import { Register, Registers } from './registers'
+import { OffsetEdgeTransformationBehaviour }      from './utils/offsetSelection'
```
Comment (71) [inline src/extension.ts]:
```suggestion
import { OffsetEdgeTransformationBehaviour } from './utils/offsetSelection'
```

File: src/extension.ts [hunk 2]
```diff
@@ -167,6 +168,19 @@ export class Extension implements vscode.Disposable {
   readonly registers = new Registers()
   readonly history   = new HistoryManager()
 
+  readonly selectionDecorationType = vscode.window.createTextEditorDecorationType(
+    <vscode.DecorationRenderOptions> {
```
Comment (71) [inline src/extension.ts]:
```suggestion
    {
```

File: src/extension.ts [hunk 3]
```diff
@@ -167,6 +168,19 @@ export class Extension implements vscode.Disposable {
...
+      color:  { id: "editor.foreground" },
+      backgroundColor:  { id: "editor.selectionBackground" }
+    }
+  )
+  
+  readonly selectionInsertDecorationType = vscode.window.createTextEditorDecorationType(
+    <vscode.DecorationRenderOptions> {
```
Comment (71) [inline src/extension.ts]:
```suggestion
    {
```

File: src/extension.ts [hunk 4]
```diff
@@ -167,6 +168,19 @@ export class Extension implements vscode.Disposable {
   readonly registers = new Registers()
   readonly history   = new HistoryManager()
 
+  readonly selectionDecorationType = vscode.window.createTextEditorDecorationType(
+    <vscode.DecorationRenderOptions> {
+      color:  { id: "editor.foreground" },
```
Comment (71) [inline src/extension.ts]: Single quotes, please (same below).

File: src/extension.ts [hunk 5]
```diff
@@ -223,6 +256,14 @@ export class Extension implements vscode.Disposable {
     return
   }
 
+  restoreLastSelections(editor: vscode.TextEditor) {
+    let documentHistory = this.history.for(editor.document)
+    let lastSels = documentHistory.getLastSelections(editor.document)
+    if (lastSels.length > 0 ) {
```
Comment (71) [inline src/extension.ts]:
```suggestion
    if (lastSels.length > 0) {
```

File: src/extension.ts [hunk 6]
```diff
@@ -387,10 +432,35 @@ export class Extension implements vscode.Disposable {
...
+              //! hence the 'last selections' in the document history has been already updated.
+              //! If new selections intersect with last selections of document changes, 
+              //! it is assumed that the last selection must not be reset.
+              let documentHistory = this.history.for(e.textEditor.document)
+              let lastSels = documentHistory.getLastSelections(e.textEditor.document)
+              if(lastSels.length === e.selections.length) {
+                reset = false
+                for (let index in e.selections) {
+                  let intersection = lastSels[index].intersection(e.selections[index])
+                  if(lastSels[index].intersection(e.selections[index]) === undefined) {
```
Comment (71) [inline src/extension.ts]:
```suggestion
                  if (lastSels[index].intersection(e.selections[index]) === undefined) {
```

File: src/extension.ts [hunk 7]
```diff
@@ -387,10 +432,35 @@ export class Extension implements vscode.Disposable {
...
+            //! If selection kind is Command it should/can be produced by insertion/append from dance
+            //! If selection kind is Keyboard it should/can be produced by textchanges
+            if(!!e.kind && ((e.kind === vscode.TextEditorSelectionChangeKind.Command) || (e.kind === vscode.TextEditorSelectionChangeKind.Keyboard))) {
+              //! Luckily document changes are fired before selection changes,
+              //! hence the 'last selections' in the document history has been already updated.
+              //! If new selections intersect with last selections of document changes, 
+              //! it is assumed that the last selection must not be reset.
+              let documentHistory = this.history.for(e.textEditor.document)
+              let lastSels = documentHistory.getLastSelections(e.textEditor.document)
+              if(lastSels.length === e.selections.length) {
```
Comment (71) [inline src/extension.ts]:
```suggestion
              if (lastSels.length === e.selections.length) {
```

File: src/extension.ts [hunk 8]
```diff
@@ -387,10 +432,35 @@ export class Extension implements vscode.Disposable {
...
           const mode = this.modeMap.get(e.textEditor.document)
 
-          if (mode === Mode.Insert)
+          if (mode === Mode.Insert) {
             this.setDecorations(e.textEditor, this.insertMode.decorationType)
-          else
+            let reset = true
+            //! If selection kind is Command it should/can be produced by insertion/append from dance
+            //! If selection kind is Keyboard it should/can be produced by textchanges
+            if(!!e.kind && ((e.kind === vscode.TextEditorSelectionChangeKind.Command) || (e.kind === vscode.TextEditorSelectionChangeKind.Keyboard))) {
```
Comment (71) [inline src/extension.ts]:
```suggestion
            if (e.kind !== undefined && ((e.kind === vscode.TextEditorSelectionChangeKind.Command) || (e.kind === vscode.TextEditorSelectionChangeKind.Keyboard))) {
```

Comment (71) [reply src/extension.ts]: I don't like implicit conversions, can you explicitly compare against the value we don't want?

File: src/extension.ts [hunk 9]
```diff
@@ -387,10 +432,35 @@ export class Extension implements vscode.Disposable {
...
+                for (let index in e.selections) {
+                  let intersection = lastSels[index].intersection(e.selections[index])
+                  if(lastSels[index].intersection(e.selections[index]) === undefined) {
+                    reset = true
+                    continue
+                  }
+                }
+              } 
+            }
+            if(reset) {
```
Comment (71) [inline src/extension.ts]:
```suggestion
            if (reset) {
```

File: src/history.ts [hunk 10]
```diff
@@ -1,6 +1,8 @@
 import * as vscode from 'vscode'
 
 import { CommandDescriptor, CommandState, InputKind } from './commands'
+import { OffsetRange, OffsetSelection, OffsetEdgeTransformationBehaviour } from './utils/offsetSelection'
```
Comment (71) [inline src/history.ts]: Please order imports alphabetically (in this line and by swapping with the next line).

File: src/history.ts [hunk 11]
```diff
@@ -29,11 +31,97 @@ export class HistoryManager {
...
+          return <PreparedChanges> {
+            initialRemoveRange: offRange,
+            remove: offRange,
+            insert: c.text.length,
+            offsetChange: c.text.length - c.rangeLength,
+            absOffsetChange: c.text.length - c.rangeLength,
+            absOffsetChangeBefore: 0,
+          }
+      })
+    for(let i=1; i < prepChanges.length; ++i) {
```
Comment (71) [inline src/history.ts]:
```suggestion
    for (let i = 1; i < prepChanges.length; i++) {
```

File: src/history.ts [hunk 12]
```diff
@@ -29,11 +31,97 @@ export class HistoryManager {
...
   readonly commands = [] as [CommandDescriptor<any>, CommandState<any>][]
   readonly changes = new WeakMap<CommandState<any>, vscode.TextDocumentContentChangeEvent[]>()
+  readonly marks = new Map<Register, OffsetSelection[]>()
+  lastSelections = [] as OffsetSelection[]
+  lastBufferModification = undefined as vscode.Range | undefined
+
+  private prepareChanges(changes: vscode.TextDocumentContentChangeEvent[]) {
+    let prepChanges = changes.slice() // Copy arry with slice
+      .sort( (c1, c2) => c1.range.start.compareTo(c2.range.start) )
+      .map( (c) => {
```
Comment (71) [inline src/history.ts]:
```suggestion
      .sort((c1, c2) => c1.range.start.compareTo(c2.range.start))
      .map(c => {
```

File: src/history.ts [hunk 13]
```diff
@@ -29,11 +31,97 @@ export class HistoryManager {
...
+   *    If there are performance issues for large files, a reimplementation is necessary.
+   *    Then TextDocumentContentChangeEvent.range information can be used to determine
+   *
+   *      * how many lines are removed for changes occuring before any selection
+   *      * special handling when changes are removed on same lines...
+   *
+   *    Similarily TextDocumentContentChangeEvent.text is used to determine the number of inserted lines by counting the occurrenc of '\n'...
+   */
+  private updateChanges(offsetSelections: OffsetSelection[], changes: PreparedChanges[]) {
+    return offsetSelections.map( s => {
```
Comment (71) [inline src/history.ts]:
```suggestion
    return offsetSelections.map(s => {
```

**Comment by geier1993** [comment]
Ok I hope most of the changes are done now.

Can we extend the tslint.json  to
`"extends": "tslint:recommended",`?

Most of the options you have requested are in there... whitespace after for/if, const instead of let.
For quotemarks we can add a rule:
 
   ```
 "quotemark": [
      true,
      "single",
      "avoid-escape",
      "avoid-template"
    ],
```

Then it is easier for me to complete you demands.
If you want to, we can create a further branch/pull request just for tslint fixes.
Then I take the time to fix all files according to a proper tslint configuration.

**Comment by Delapouite** [comment]
On a side note, TSLint is now deprecated and should be replaced by ESLint (which is now able to parse TypeScript): https://medium.com/palantir/tslint-in-2019-1a144c2317a9

**Comment by 71** [comment]
@geier1993 @Delapouite Good idea, thank you! Sorry, I did add ESLint and fix the rules, so it might make it harder to rebase. I'll look at the changes tomorrow.

FYI I really want this feature but I've been prioritizing fixing bugs in hope of "keeping the best for the end."

**Comment by geier1993** [comment]
Yay sounds great.
Do you mind if I merge the updates in my feature branch instead of rebasing something? ... Some more commits but it makes it easier for me now.

I think I have already merged with you master in the past without thinking about it it

**Comment by 71** [comment]
@geier1993 Sure, whatever makes it easier. We'll edit the history later if we have to.

I do have some bad news. After working on getting rid of the bugs introduced with `allowEmpty: false`, I came up with a new solution that provides a new layer on top of VS Code selections and that kind of makes your `OffsetSelection` useless (it started as a slightly different interface which handled single-cursor selections and added caching / faster operations, but eventually evolved into selections that are tracked and updated on document change, like yours. I was able to implement `saveSelections` / `restoreSelections` in basically a single statement after that).

I'm planning on rewriting a lot of stuff to use that API, but that will make a lot of your changes obsolete. I feel terribly sorry that you spent so much time on this CL to eventually have it become (mostly) useless. While many of the changes you made will still be useful after my changes, it does mean that they will have to be changed to adhere to the new API.

If you'd like to see that PR to completion, then please wait until the changes have been added and update everything. Otherwise, I'll take care of it myself (but still credit you, since some of the changes were inspired by the commit, and since some of your work will have to be reused anyway).

**Comment by geier1993** [comment]
Ok no problem, just let me know when I can help you :)

**Comment by 71** [comment]
This was just implemented in `unstable`: 1a038e74bbdeb3240ed29c053fb3f090624c63aa.

If you notice any problem with it, please let me know!

**Pull request closed without merging**